### PR TITLE
google oauth setup

### PR DESCRIPTION
**Pull request by Spandana-Baruah**
google oauth setup without the database integration for now. Will work on that as the next step from this. Wanted to get this up and running along with the fb oauth so that we can proceed with other stuff.

### CODE REVIEW

File: api/app.js [hunk 1]
```diff
@@ -2,42 +2,134 @@ var createError = require('http-errors');
...
+const passportLocalMongoose = require('passport-local-mongoose');
+const googleSetup = require("./google-setup");
+const findOrCreate = require('mongoose-findorcreate');
 
 var app = express();
 
+app.use(
+  cookieSession({
+    name: "session",
+    keys: "thisappisawesome",
```
Comment (aaronNgu) [inline api/app.js]: It should be `keys: ["thisappisawesome"]`.  This was the one that was causing the error.

File: api/google-setup.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+const passport = require('passport');
+const GoogleStrategy = require('passport-google-oauth20').Strategy;
+
+
+
+passport.use(new GoogleStrategy({
+    clientID: "798089317931-hca1vqdbgf41bcu3vnbsg4c7f1h58gdu.apps.googleusercontent.com",
+    clientSecret: "jQCNTrDmsy9sTwy8ANjtxCwB",
```
Comment (aaronNgu) [inline api/google-setup.js]: This should probably be in the .env file

File: api/google-setup.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    callbackURL: "http://localhost:3000/homepage",
```
Comment (aaronNgu) [inline api/google-setup.js]: I think the callbackURL should be `/auth/google/redirect` because passport.authenticate needs to be called again. Just found out about it too. [here](https://stackoverflow.com/questions/26268271/why-is-passport-authenticate-called-twice) I think this is probably why the setup isn't quite working properly.

File: api/routes/auth-routes.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      isAuthenticated: false,
+      message: "user has not been authenticated"
+    });
+  } else {
+    next();
+  }
+};
+
+router.get("/checkStatus", authCheck, (req, res) => {
+    passport.authenticate('google', { scope: ["profile"] }),
```
Comment (aaronNgu) [inline api/routes/auth-routes.js]: I think at this point we don't need `passport.authenticate()` anymore

Comment (aaronNgu) [reply api/routes/auth-routes.js]: `/checkStatus` isn't quite working properly. I dont know if its because the authentication wasn't completed or because theres a bug with route itself

File: api/app.js [hunk 5]
```diff
@@ -2,42 +2,134 @@ var createError = require('http-errors');
...
+var port = 9000;
 
 var indexRouter = require('./routes/index');
 var usersRouter = require('./routes/users');
 var coursesRouter = require('./routes/courses');
 var reviewsRouter = require('./routes/reviews');
+var authRouter = require('./routes/auth-routes');
+
+const cookieSession = require("cookie-session");
+const session = require('express-session');
```
Comment (aaronNgu) [inline api/app.js]: don't need this anymore

**Review by aaronNgu — COMMENTED** [review]
Added some comments. I'm not entirely sure why it isn't working properly. I'm going to try to build the Facebook one to see if I can find out where.

File: api/routes/auth-routes.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      message: "user has successfully authenticated",
+      user: req.user,
+      cookies: req.cookies
+    });
+  }
+});
+
+const authCheck = (req, res, next) => {
+  if (!req.user) {
+    res.status(401).json({
```
Comment (aaronNgu) [inline api/routes/auth-routes.js]: Just an observation. After logging in, it still comes to this part of the code when I call `http://localhost:9000/auth/checkStatus` which implies that req.user is null. But it shouldn't be if I have logged in successfully

**Review by aaronNgu — COMMENTED** [review]
Is it possible to remove the changes made in the /LoginPage.js and /LoginPageBoc.js?

**Review by aaronNgu — COMMENTED** [review]
Also the passport and prop-types in react_app/package.json is no needed

**Review by wwyc — CHANGES_REQUESTED** [review]
should probably keep .env-example.txt file as a reference?

File: api/app.js [hunk 7]
```diff
@@ -38,14 +72,15 @@ app.use('/', indexRouter);
...
+app.use("/auth", authRouter);
 
 // catch 404 and forward to error handler
-app.use(function(req, res, next) {
+app.use(function (req, res, next) {
   next(createError(404));
 });
 
 // error handler
-app.use(function(err, req, res, next) {
```
Comment (wwyc) [inline api/app.js:48]: adding extra space not needed?

File: api/app.js [hunk 8]
```diff
@@ -10,24 +10,58 @@ var indexRouter = require('./routes/index');
 var usersRouter = require('./routes/users');
 var coursesRouter = require('./routes/courses');
 var reviewsRouter = require('./routes/reviews');
+var authRouter = require('./routes/auth-routes');
+
+const cookieSession = require("cookie-session");
+const passport = require('passport');
+const googleSetup = require("./google-setup");
```
Comment (wwyc) [inline api/app.js]: is this required?

File: api/app.js [hunk 9]
```diff
@@ -38,14 +72,15 @@ app.use('/', indexRouter);
...
 
 // catch 404 and forward to error handler
-app.use(function(req, res, next) {
+app.use(function (req, res, next) {
   next(createError(404));
 });
 
 // error handler
-app.use(function(err, req, res, next) {
+app.use(function (err, req, res, next) {
```
Comment (wwyc) [inline api/app.js]: adding extra space not neeed?

File: api/app.js [hunk 10]
```diff
@@ -38,14 +72,15 @@ app.use('/', indexRouter);
 app.use('/users', usersRouter);
 app.use('/courses', coursesRouter);
 app.use('/reviews', reviewsRouter);
+app.use("/auth", authRouter);
 
 // catch 404 and forward to error handler
-app.use(function(req, res, next) {
+app.use(function (req, res, next) {
```
Comment (wwyc) [inline api/app.js]: adding extra space not neeed?

File: api/routes/auth-routes.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
+const router = require("express").Router();
+const passport = require("passport");
+const CLIENT_HOME_PAGE_URL = "http://localhost:3000/homepage";
```
Comment (wwyc) [inline api/routes/auth-routes.js:3]: hard code to local host 3000 may cause deployment issues

File: api/app.js [hunk 12]
```diff
@@ -10,24 +10,58 @@ var indexRouter = require('./routes/index');
...
+// parse cookies
+app.use(cookieParser());
+
+app.use(passport.initialize());
+app.use(passport.session());
+
 //connect to mongo db
 const mongoose = require('mongoose');
 const config = process.env;
-mongoose.connect('mongodb+srv://' + config.DB_USER  + ':' + config.DB_PW + '@sandbox-7vuqw.mongodb.net/' + config.DB_DBNAME + '?retryWrites=true&w=majority')
```
Comment (wwyc) [inline api/app.js:19]: extra spaces not needed?

File: api/app.js [hunk 13]
```diff
@@ -10,24 +10,58 @@ var indexRouter = require('./routes/index');
...
 
 // view engine setup
 app.set('views', path.join(__dirname, 'views'));
 app.set('view engine', 'jade');
 
-app.use(cors());
+// set up cors to allow us to accept requests from our client
+app.use(
+  cors({
+    origin: "http://localhost:3000", // allow to server to accept request from different origin
```
Comment (wwyc) [inline api/app.js:62]: hard code to local host 3000 may cause deployment issues

File: api/app.js [hunk 14]
```diff
@@ -55,4 +147,7 @@ app.use(function(err, req, res, next) {
   res.render('error');
 });
 
+// connect react to nodejs express server
```
Comment (wwyc) [inline api/app.js]: remove commented out code?

File: api/app.js (hunk 9 above)
Comment (aaronNgu) [reply api/app.js]: my bad that's my linter

File: api/app.js [hunk 15]
```diff
@@ -10,24 +10,58 @@ var indexRouter = require('./routes/index');
...
+var authRouter = require('./routes/auth-routes');
+
+const cookieSession = require("cookie-session");
+const passport = require('passport');
+require("./google-setup");
 
 var app = express();
 
+app.use(
+  cookieSession({
```
Comment (wwyc) [inline api/app.js:25]: how are the cookie session object items defined?

Comment (aaronNgu) [reply api/app.js:25]: Do you mean how is the cookie defined/set? They are handled by passport but we can specify what gets put in the cookie. If you look at line 44, we have a `passport.serializeUser()`, that function allows us to decide what to put in the cookie. In this case, we're passing in the whole user object (from the parameter) into the cookie. We don't get to call the `passport.serializeUser()`. I am not entirely sure but I think the user object(from the parameter) is the user object from the Facebook/Google OAuth response.

Comment (wwyc) [reply api/app.js:25]: Interesting!  I always wondered how cookies are handled.  I was also wondering how the items name, keys, maxAge is used?

Comment (aaronNgu) [reply api/app.js:25]: I'm not entirely sure what the name and keys do. but I think the maxAge defines how long the cookie is valid

**Review by aaronNgu — APPROVED** [review]

File: api/app.js [hunk 16]
```diff
@@ -10,24 +10,61 @@ var indexRouter = require('./routes/index');
 var usersRouter = require('./routes/users');
 var coursesRouter = require('./routes/courses');
 var reviewsRouter = require('./routes/reviews');
+var authRouter = require('./routes/auth-routes');
+
+const cookieSession = require("cookie-session");
+const passport = require('passport');
+require("./google-setup");
+
+// CITATION: used https://medium.com/free-code-camp/how-to-set-up-twitter-oauth-using-passport-js-and-reactjs-9ffa6f49ef0
```
Comment (wwyc) [inline api/app.js:19]: I think this was already added to read me?

File: react_app/src/components/LoginPage/LoginPageBox.js [hunk 17]
```diff
@@ -13,7 +13,7 @@ const useStyles = makeStyles({
 
 const LoginPageBox = (props) => {
     const classes = useStyles();
-
+    
     return (
```
Comment (wwyc) [inline react_app/src/components/LoginPage/LoginPageBox.js:17]: Looks like no changes were made to this file? Pls revert the changes before merging.  Thanks.

File: react_app/package.json [hunk 18]
```diff
@@ -8,6 +8,8 @@
     "@testing-library/jest-dom": "^4.2.4",
     "@testing-library/react": "^9.3.2",
     "@testing-library/user-event": "^7.1.2",
+    "passport": "^0.4.1",
+    "prop-types": "^15.7.2",
```
Comment (wwyc) [inline react_app/package.json]: is prop-types needed?

**Review by wwyc — APPROVED** [review]

**Comment by wwyc** [comment]
See new comments added.  Otherwise looks good!

**Review by wwyc — APPROVED** [review]
See new comments and changes requested.  
As discussed the hard coded local hosts wont work for deployment but @aaronNgu will fix it in future PRs.
Otherwise looks good!

File: react_app/package.json (hunk 18 above)
Comment (Spandana-Baruah) [reply react_app/package.json]: Not for now but it will be for the next DB integration PR - I'll still remove it for this PR though to keep things clean.

File: react_app/src/components/LoginPage/LoginPageBox.js (hunk 17 above)
Comment (Spandana-Baruah) [reply react_app/src/components/LoginPage/LoginPageBox.js:17]: Yeah, we don't currently need to make changes to this file for the purposes of this PR.

File: api/app.js (hunk 16 above)
Comment (Spandana-Baruah) [reply api/app.js:19]: We still need to be adding citations to the local code as well. If you look at the syllabus here https://blogs.ubc.ca/cpsc436i2020s/syllabus/#sec-5-5 "Citations/acknowledgments should be in a clear section in your main README.md, in your license (if you have one!), and repeated locally where you used the code/help."
